src/odemis/driver/andorshrk.py

```diff
@@ -333,27 +358,27 @@ def __init__(self, name, role, device, camera=None, accessory=None,
         self._slit_names.update(slits)
 
         self.Initialize()
+        self._reconnecting = False
+
         try:
-            if isinstance(device, basestring):
-                self._device = self._findDevice(device)
-            else:
-                nd = self.GetNumberDevices()
-                if device >= nd:
-                    raise HwError("Failed to find Andor Shamrock (%s) as device %s" %
-                                  (name, device))
-                self._device = device
-
-            # TODO: EEPROM contains name of the device, but there doesn't seem to be any function for getting it?!
-            fl, ad, ft = self.EepromGetOpticalParams()
-            if 0.190 <= fl <= 0.200:
-                self._model = MODEL_KY193
-            elif 0.296 <= fl <= 0.310:
-                self._model = MODEL_SR303
-            elif 0.326 <= fl <= 0.330:
-                self._model = MODEL_KY328
-            else:
-                self._model = None
-                logging.warning("Untested spectrograph with focus length %d mm", fl * 1000)
+            self._model, fl = self._get_model(name, device)
+            if not self._model:
+                # If the spectrograph is in a weird state (e.g. due to em interference), it might
+                # not be detected, try to power cycle and reconnect.
+                logging.info("Unrecognized spectrograph model, trying to reconnect.")
+                # For power cycling, we need the ._power_supplier attribute which has not been
+                # initialized yet. Initializing it now if it's there.
+                if 'power_supplier' in kwargs:
+                    self._power_supplier = kwargs['power_supplier']
+                    self._reconnect(on_startup=True)
+                else:
+                    # Hope everything is fine and the spectrograph just has a weird model number
+                    logging.info("Cannot reconnect because spectrograph doesn't have a power supplier.")
+                # Try again, if we still don't recognize the model, the spectrograph is probably fine and
+                # we just don't recognize the model number.
+                self._model, fl = self._get_model(name, device)
+                if not self._model:
+                    logging.warning("Untested spectrograph with focus length %d mm", fl * 1000)
```

Comment:
I'd rather avoid extra turning on/off at startup. If it's messed-up at start-up, let the user handle it. There might be a lot of other reason than this issue.